### PR TITLE
Changed the publisher lifecycle data retrieving process to JSON format

### PR DESCRIPTION
**Pull request by shnrndk**
Fixes wso2/api-manager#314

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        LCManager.tenantDomain = tenantDomain;
+    }
+    public static String getTenantDomain() {
+        return tenantDomain;
+    }
+
+    public LCManager(String tenantDomain) throws APIManagementException
+    {
+        setTenantDomain(tenantDomain);
+    }
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java:54]: Seems this class is not formatted

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java [hunk 2]
```diff
@@ -0,0 +1,305 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java:2]: Invalid header

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    public static String getTenantDomain() {
+        return tenantDomain;
+    }
+
+    public LCManager(String tenantDomain) throws APIManagementException
+    {
+        setTenantDomain(tenantDomain);
+    }
+
+    public void processLifeCycle() throws APIManagementException {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java:56]: this is an internal method. make it private

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManagerFactory.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManagerFactory.java:2]: Add correct header

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    private static JSONObject defaultLCObj;
+    private LCManager instance;
+
+    static {
+        if (defaultLCObj == null) {
+
+            try {
+                defaultLCObj = getDefaultLCConfigJSON();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java:71]: any reason to handle this in static block

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java (hunk 5 above)
Comment (shnrndk) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java:71]: This was done to reduce the number of file readings. Due to this Static code block the file reading is only done at the start of the program. If we didn't use this static block, this getDefaultLCConfigJSON() would execute on each state change, causing the file readings to occur on each state change, which happens to be the worst case scenario.

**Review by tharindu1st — APPROVED** [review]

**Comment by rmsamitha** [comment]
Better to catch separate exceptions thrown by different code lines, separately.

**Comment by shnrndk** [comment]
> components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java Line 2796 : Better to catch separate exceptions thrown by different code lines, separately.

I have checked this, This exception is removed currently in the APIProviderImpl.java in the changeLifeCycleStatus() method in master.

**Pull request merged**